src/Chapter4.hs

```diff
@@ -489,6 +505,14 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+instance Applicative (List) where
+    pure :: a -> List a
+    pure x = Cons x Empty
+
+    (<*>) :: List (a -> b) -> List a -> List b
+    (<*>) Empty _ = Empty
+    (<*>) _ Empty = Empty
+    (<*>) (Cons f fs) (Cons x xs) = Cons (f x) (fs <*> xs)
```

Comment:
There are two possible ways to implement the `Applicative` instance for the list, you choose the one where each function is only applied to the corresponding one element from the values list. 
Another way to implement it is to apply each function to each element.

Unfortunately, in Haskell, it's not possible to implement two different instances of the same typeclass for the same data type, so you must choose one. 

For list, it was chosen so the laws of `Monad` and `Applicative` typeclasses agree with each other. Because `Applicative` is a superclass for `Monad` there are some laws (implicit contract) that tells how functions must behave. For the `Applicative-Monad` relationship, the following must hold:

```haskell
mf <*> mx ≡ mf >>= (\f -> mx >>= (\x -> return (f x)))
```

If you implement the `Applicative` instance the way you did here, it's not possible to implement the `Monad` instance to satisfy this rule. 

However, it doesn't mean that such an instance is invalid. When you want to provide multiple instances in Haskell, you use newtypes. And there's the `ZipList` newtype that provides the described `Applicative` instance, and doesn't have the `Monad` instance as a consequence:

* https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Applicative.html#t:ZipList